cirq-google/cirq_google/workflow/quantum_executable.py

```diff
@@ -0,0 +1,157 @@
+import abc
+import dataclasses
+from dataclasses import dataclass
+from typing import Union, Tuple, Optional, Sequence, cast
+
+import cirq
+from cirq import NamedTopology, _compat
+from cirq.protocols import dataclass_json_dict
+from cirq.study.resolver import _is_param_resolver_or_similar_type
+
+
+class ExecutableSpec(metaclass=abc.ABCMeta):
+    """Specification metadata about an executable.
+
+    Subclasses should add problem-specific fields.
+    """
+
+    executable_family: str = NotImplemented
+    """A unique name to group executables."""
+
+
+@dataclass(frozen=True)
+class BitstringsMeasurement:
+    """Use in-circuit MeasurementGate to collect many repetitions of strings of bits.
+
+    This is the lowest-level measurement type allowed in `QuantumExecutable` and behaves
+    identically to the `cirq.Sampler.run` function. The executable's circuit must contain
+    explicit measurement gates.
+
+    Args:
+        n_repeitions: The number of repetitions to execute the circuit.
+    """
+
+    n_repetitions: int
+
+    def _json_dict_(self):
+        return dataclass_json_dict(self, namespace='cirq.google')
+
+    def __repr__(self):
+        return _compat.dataclass_repr(self, namespace='cirq_google')
+
+
+TParamPair = Tuple[cirq.TParamKey, cirq.TParamVal]
+
+
+@dataclass(frozen=True)
+class QuantumExecutable:
+    """An executable quantum program.
+
+    This serves a similar purpose to `cirq.Circuit` with some key differences. First, a quantum
+    executable contains all the relevant context for execution including parameters as well as
+    the desired number of repetitions. Second, this object is immutable. Finally, there are
+    optional fields enabling a higher level of abstraction for certain aspects of the executable.
+
+    Attributes:
+        circuit: A circuit describing the quantum operations to execute.
```

Comment:
Nit: A \`cirq.Circuit\` describing.... so the docs linking can happen.

Reply:
done, but isn't this redundant with the type annotations?

Reply:
It is, but when rendered to quantumai.google this allows the docstrings to generate hyperlinks on the webpages which is nice. Historically we haven't been doing this very diligently  (function signatures in the code sometimes do this, but I find it to be a little less reliable). https://www.tensorflow.org/api_docs/python/tf/math/ceil (See the orange args box that has the hyperlink to the `tf.Tensor`)

Reply:
Looks good. It might make sense to broadcast to the devs that we're going to start being more diligent about this and/or put it in the style/contributing guide.